### ISSUE TITLE
LAN access in VPN Lockdown mode

### ISSUE THREAD

**Issue by mvevitsis** [celzero/rethink-app#1618]
Continuing the discussion from reddit:

System firewall (lockdown mode):
Outgoing to LAN -> blocked
Incoming from LAN -> allowed

RethinkDNS + system firewall (lockdown mode)
Outgoing to LAN -> allowed, unless you exclude the app in which case it will be blocked by the system firewall as above (can also be blocked directly by rethink firewall rules, e.g. block 192.168.1.*:5353) 
Incoming from LAN -> not working (routing issue?)

So basically:
To send, you need rethink on. To receive, you need rethink stopped (not paused!)

Strangely, netguard has the opposite issue. With netguard on and in lockdown mode you can receive, but not send.

I haven't looked at netguard's source code yet (or rethink's) to compare them.

**Comment by mvevitsis** [comment]
Update:

Invizible handles both outgoing and incoming LAN connections properly, even with lockdown mode enabled. 

https://github.com/Gedsh/InviZible

I haven't looked at the source yet but this proves it is possible to fix.

**Comment by ignoramous** [comment]
A user [writes](https://www.reddit.com/r/rethinkdns/comments/1dm6oyn/comment/lf4w3sk/) ([mirror](https://archive.is/GoNxE)),

>  To reproduce, enable rethink, enable lockdown mode (always on VPN + block connections without VPN) then make sure you do not enable do not route LAN thru rethink.
>
> Open local send on the computer and phone.
>
> Confirm you can send a file from the phone to the computer.
>
> Now try sending from the computer to the phone, confirm the computer cannot find the phone.
>
> Next enable 'do not route... ' in rethink and repeat the steps (only possible on older versions of rethink, now if you enable lockdown this option gets greyed out).
>
> You will notice this time the phone cannot send to the computer, but the computer can send to the phone (this is expected because block connections without VPN only blocks incoming connections).

**Comment by ignoramous** [comment]
>  Incoming from LAN -> not working (routing issue?)

I think Rethink wouldn't let *Incoming from LAN* with or without *VPN Lockdown* (ie, *Block connections without VPN*) turned ON. Can you confirm? If so, this problem won't be solved unless we overhaul how we treat incoming packets (which is not a trivial undertaking and as such I don't see us investing effort in to it anytime soon given the already existing open feature requests / issues).

**Comment by mvevitsis** [comment]
That's correct. Even with block connections without VPN off, as long as they are being routed thru rethink incoming LAN connections do not work.

**Comment by ignoramous** [comment]
This issue then is a dup of #577 (closing, but feel free to re-open)

**Comment by ignoramous** [comment]
Track impl at: https://github.com/celzero/firestack/issues/77

**Comment by mvevitsis** [comment]
> Track impl at: https://github.com/celzero/firestack/issues/77

Bookmarked, looking forward to possible fix.

**Comment by Lanius-collaris** [comment]
Endpoint-Independent Mapping is for UDP punching, can't fix accepting connections initiated by other devices in a LAN.
https://tailscale.com/blog/how-nat-traversal-works 

---

> Incoming from LAN -> not working (routing issue?)

I think this is expected behavior if your subnet is routed via the TUN controlled by rethink.
You can check with
`adb shell ip route get <IP>` 

Let rethink run a reverse proxy may be a solution. @ignoramous

**Comment by mvevitsis** [comment]
InviZible Pro is able to route incoming LAN requests. 

Checking it's routing behavior via adbshell, I used the command in your post above to check 10.0.0.1

The result is that it is routed over tun1 which is indeed the interface currently in use by inviZible.

**Comment by mvevitsis** [comment]
InviZible Pro is able to route incoming LAN requests. 

Checking it's routing behavior via adbshell, I used the command in your post above to check 10.0.0.1

The result is that it is routed over tun1 which is indeed the interface currently in use by inviZible.

Of course invizible doesn't have a built in wireguard client. I was able to get it to work, but I don't consider this doable for most people:

![Messenger_creation_f1113f24-9752-4288-bb1e-3ef7d5605dba.jpeg](https://github.com/user-attachments/assets/b7e7f0b6-912e-460f-90ac-23248a921bde)

**Comment by ignoramous** [comment]
> Endpoint-Independent Mapping is for UDP punching, can't fix accepting connections initiated by other devices in a LAN.
https://tailscale.com/blog/how-nat-traversal-works

If Rethink allows UDP hole-punching, wouldn't those P2P apps be able to connect through to the endpoints on this side of (a full-cone NAT?) Rethink?

**Comment by Lanius-collaris** [comment]
Interesting, [Intra](https://github.com/Jigsaw-Code/Intra) allows incoming TCP as well as InviZible Pro.

**Comment by mvevitsis** [comment]
InviZible Pro is able to route incoming LAN requests. 

Checking it's routing behavior via adbshell, I used the command in your post above to check 10.0.0.1

The result is that it is routed over tun1 which is indeed the interface currently in use by inviZible.

Of course invizible doesn't have a built in wireguard client. I was able to get it to work, but I don't consider this doable for most people:

![Messenger_creation_f1113f24-9752-4288-bb1e-3ef7d5605dba.jpeg](https://github.com/user-attachments/assets/b7e7f0b6-912e-460f-90ac-23248a921bde)


Tethering is allowed by the system even with lockdown mode on; I don't think these are related. 

I could ask him but he probably doesn't't like me because I argued with him about letting me set the bootstrap resolvers for dnscrypt to 127.0.0.1 or null (I use a stamp, so I don't need them).

**Comment by ignoramous** [comment]
> Interesting, [Intra](https://github.com/Jigsaw-Code/Intra) allows incoming TCP as well as InviZible Pro.

Intra *allows* apps to bind to underlying interfaces. I sent a PR to fix this, but they didn't bother (and so, Intra should *not* even work when VPN is in *Lockdown* mode aka *Block connections without VPN* is turned ON): https://github.com/Jigsaw-Code/Intra/pull/420

Intra uses tun2socks wrapper over LwIP (and supports endpoint-independent mapping), but Rethink has switched to gvisor's netstack.

**Comment by Lanius-collaris** [comment]
> Intra _allows_ apps to bind to underlying interfaces.

Seems Invizible has the same bug. Tested with `busybox nc -l 10000 -p -s ::%30` (30 is my wlan0) @ignoramous

**Comment by mvevitsis** [comment]
> > Intra _allows_ apps to bind to underlying interfaces.
> 
> Seems Invizible has the same bug. Tested with `busybox nc -l -p 10000 -s ::%30` (30 is my wlan0) @ignoramous

I am not rooted, so I can't use BusyBox, but I haven't seen any apps fail to connect after enabling lockdown mode with invizible. 

However, much like rethink's 'use all available connections' option, using invizible seems to ignore Android's active network (both WWAN and WLAN icons are showing in the status bar).

**Comment by ignoramous** [comment]
> Seems Invizible has the same bug.

It isn't a bug? From what I know, the VPN app has to *opt-in* to allow sockets from other installed apps to bind to interfaces other than the VPN's tun device. From its code, InviZible does not opt-in to allow for this. The bind may be working because InviZible excludes IP certain IP ranges by default (as discussed above).

>  much like rethink's 'use all available connections' option, using invizible seems to ignore Android's active network (both WWAN and WLAN icons are showing in the status bar). I am not sure why it is doing that.

Why is this concerning? If "use all available networks" is turned ON, then the VPN app, if it supports it, must be able to connect via any of the active networks (wifi, mobile, ethernet, usb etc)

**Comment by mvevitsis** [comment]
Not concerning, just thought it was interesting. There isn't a setting for it as far as I can tell.

**Comment by Lanius-collaris** [comment]
> It isn't a bug?

Sorry. Doesn't appear when tor is enabled (Some very old versions do?), so it's not a bug.
@ignoramous

**Comment by Lanius-collaris** [comment]
@mvevitsis 
> I am not rooted, so I can't use BusyBox

Running Busybox doesn't require a `/system/bin/busybox`, you can install a terminal emulator (e.g. Termux) or place a statically linked BusyBox in `/data/local/tmp/`.

**Comment by mvevitsis** [comment]
> @mvevitsis 
> > I am not rooted, so I can't use BusyBox
> 
> Running Busybox doesn't require a `/system/bin/busybox`, you can install a terminal emulator (e.g. Termux) or place a statically linked BusyBox in `/data/local/tmp/`.

I got it working in termux, thanka

**Comment by ignoramous** [comment]
> InviZible Pro

Apparently, InviZible working for inbound LAN is due to broken *Block connections without VPN* (VPN Lockdown) implementation in Android: https://issuetracker.google.com/issues/280462382 / via: https://github.com/mullvad/mullvadvpn-app/issues/4643#issuecomment-1531523100

**Comment by mvevitsis** [comment]
> > InviZible Pro
> 
> Apparently, InviZible working for inbound LAN is due to broken *Block connections without VPN* (VPN Lockdown) implementation in Android: https://issuetracker.google.com/issues/280462382 / via: https://github.com/mullvad/mullvadvpn-app/issues/4643#issuecomment-1531523100

I'm not sure it's actually broken or if it's by design. Incoming connections have never been filtered by that setting, only outgoing.
